File: ocaml/06.ml
(* 6 : Find out whether a list is a palindrome : A palindrome can be read forward or backward; e.g : (x a m a x)  *)

let rec palindome (s : string) : bool =
  if String.length s <= 1 then true
  else
    if s.[0] == s.[String.length s - 1] then 
      palindome (String.sub s 1 (String.length s - 2))
    else
      false
                   
    

File: haskell/06.hs
{-
Problem 6

(*) Find out whether a list is a palindrome. A palindrome can be read forward or backward; e.g. (x a m a x).

Example in Haskell:

*Main> isPalindrome [1,2,3]
False
*Main> isPalindrome "madamimadam"
True
*Main> isPalindrome [1,2,4,8,16,8,4,2,1]
True

-}

isPalindrome :: Eq a => [a] -> Bool
isPalindrome x = if length x <= 1 then True
                 else let theFirst = head x
                          theFinal = last x
                      in
                       if theFirst == theFinal then
                         isPalindrome $ tail $ init x
                       else
                         False
